### PR TITLE
Add arrow datafusion python dependency

### PR DESCRIPTION
**Pull request by jdye64**
Closed with #1084

### CODE REVIEW

**Comment by jdye64** [comment]
Waiting on https://github.com/oxidecomputer/typify/pull/221 to merge so that conda builds can work with new setup.

File: .github/actions/setup-builder/action.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    required: true
+    default: 'stable'
+runs:
+  using: "composite"
+  steps:
+    - name: Install Build Dependencies
+      shell: bash
+      run: |
+        apt-get update
+        apt-get install -y protobuf-compiler
```
Comment (charlesbluca) [inline .github/actions/setup-builder/action.yaml:15]: Looks like we need to `sudo` these installs:

```suggestion
        sudo apt-get update
        sudo apt-get install -y protobuf-compiler
```

https://stackoverflow.com/questions/57982945/how-to-apt-get-install-in-a-github-actions-workflow

File: dask_sql/mappings.py [hunk 2]
```diff
@@ -177,6 +178,12 @@ def sql_to_python_value(sql_type: "SqlTypeName", literal_value: Any) -> Any:
         # Calcite will always convert INTERVAL types except YEAR, QUATER, MONTH to milliseconds
         # Issue: if sql_type is INTERVAL MICROSECOND, and value <= 1000, literal_value will be rounded to 0
         return np.timedelta64(literal_value, "ms")
+    elif sql_type == SqlTypeName.INTERVAL_MONTH_DAY_NANOSECOND:
+        # DataFusion assumes 30 days per month. Therefore we multiply number of months by 30 and add to days
+        return pd.Timedelta(
+            np.timedelta64((literal_value[0] * 30) + literal_value[1], "D")
+            + np.timedelta64(literal_value[2], "ns")
+        )
```
Comment (ayushdg) [inline dask_sql/mappings.py:186]:
```suggestion
        return np.timedelta64(
            (literal_value[0] * 30) + literal_value[1], "D"
        ) + np.timedelta64(literal_value[2], "ns")
```

File: .github/workflows/conda.yml [hunk 3]
```diff
@@ -49,6 +49,11 @@ jobs:
           which python
           pip list
           mamba list
+      - name: Install Protoc
+        uses: arduino/setup-protoc@v1
+        with:
+          version: '3.x'
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (charlesbluca) [inline .github/workflows/conda.yml:56]: We should probably remove this and instead add `libprotobuf=3` to the nightly recipe's build requirements (in a similar fashion to [what's being done in ADP](https://github.com/apache/arrow-datafusion-python/blob/af1a8ebade3d537ce24f6f01a037a56177a917fa/conda/recipes/meta.yaml#L40)) so that compilers are handled entirely through conda-forge once this gets pushed into a stable release

```suggestion
```

File: .github/workflows/rust.yml [hunk 4]
```diff
@@ -30,6 +30,11 @@ jobs:
         id: detect-trigger
         with:
           keyword: "[test-df-upstream]"
+      - name: Install Protoc
+        uses: arduino/setup-protoc@v1
+        with:
+          version: '3.x'
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (charlesbluca) [inline .github/workflows/rust.yml:37]: Don't think this should be necessary since this job isn't doing any compilation

```suggestion
```

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -60,6 +60,11 @@ jobs:
         with:
           workspaces: dask_planner
           shared-key: test
+      - name: Install Protoc
+        uses: arduino/setup-protoc@v1
+        with:
+          version: '3.x'
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (charlesbluca) [inline .github/workflows/test.yml:67]: Would remove this and instead add `libprotobuf=3` to the conda dev environments so that users don't need to install it manually

```suggestion
```

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -106,6 +111,11 @@ jobs:
...
```
Comment (charlesbluca) [inline .github/workflows/test.yml:118]: Same as above

```suggestion
```

File: dask_sql/mappings.py [hunk 7]
```diff
@@ -177,6 +178,11 @@ def sql_to_python_value(sql_type: "SqlTypeName", literal_value: Any) -> Any:
         # Calcite will always convert INTERVAL types except YEAR, QUATER, MONTH to milliseconds
         # Issue: if sql_type is INTERVAL MICROSECOND, and value <= 1000, literal_value will be rounded to 0
         return np.timedelta64(literal_value, "ms")
+    elif sql_type == SqlTypeName.INTERVAL_MONTH_DAY_NANOSECOND:
+        # DataFusion assumes 30 days per month. Therefore we multiply number of months by 30 and add to days
+        return np.timedelta64(
+            (literal_value[0] * 30) + literal_value[1], "D"
+        ) + np.timedelta64(literal_value[2], "ns")
```
Comment (charlesbluca) [inline dask_sql/mappings.py:185]: This fix makes sense to me - two things I'm interested in:

- is there any way to hit the cases for `INTERVAL` and `INTERVAL_DAY` directly above this? Wasn't able to reach them with our standard interval tests, interested in what CI would look like if we remove them (and maybe also associated plumbing) altogether
- with this change, we might want to consider modifying @sarahyurick's handling for months/quarters in `DatetimeSubOperation` to make the same assumption around a uniform 30 day month for sake of consistency

File: dask_sql/physical/rex/core/literal.py [hunk 8]
```diff
@@ -161,6 +161,9 @@ def convert(
         elif literal_type == "IntervalDayTime":
             literal_type = SqlTypeName.INTERVAL_DAY
             literal_value = rex.getIntervalDayTimeValue()
+        elif literal_type == "IntervalMonthDayNano":
+            literal_type = SqlTypeName.INTERVAL_MONTH_DAY_NANOSECOND
+            literal_value = rex.getIntervalMonthDayNanoValue()
```
Comment (charlesbluca) [inline dask_sql/physical/rex/core/literal.py:166]: Same question around the `IntervalDayTime` case being reached as above

File: tests/integration/test_groupby.py [hunk 9]
```diff
@@ -217,6 +217,7 @@ def test_aggregations(c):
     GROUP BY user_id
     """
     )
+
```
Comment (charlesbluca) [inline tests/integration/test_groupby.py:220]: Just removing seemingly unnecessary diff

```suggestion
```

File: tests/integration/test_select.py [hunk 10]
```diff
@@ -95,6 +95,8 @@ def test_select_of_select_with_casing(c, df):
         """
     )
 
+    print(f"Result DF:\n{result_df.head()}")
+
```
Comment (charlesbluca) [inline tests/integration/test_select.py:99]:
```suggestion
```

**Review by charlesbluca — COMMENTED** [review]
Thanks @jdye64 🎉 a few comments around how we include `protoc` in CI/packaging and some follow up around the timedelta changes

File: dask_planner/src/sql/optimizer.rs [hunk 11]
```diff
@@ -45,8 +38,9 @@ impl DaskSqlOptimizer {
     /// optimizers as well as any custom `OptimizerRule` trait impls that might be desired.
     pub fn new() -> Self {
         debug!("Creating new instance of DaskSqlOptimizer");
+
         let rules: Vec<Arc<dyn OptimizerRule + Sync + Send>> = vec![
-            Arc::new(TypeCoercion::new()),
+            // Arc::new(TypeCoercion::new()),
```
Comment (ayushdg) [inline dask_planner/src/sql/optimizer.rs:43]: Was there a specific issue with this rule enabled?

**Review by ayushdg — CHANGES_REQUESTED** [review]

File: dask_planner/src/sql/logical.rs [hunk 12]
```diff
@@ -325,7 +327,8 @@ impl PyLogicalPlan {
             LogicalPlan::CreateCatalogSchema(_create) => "CreateCatalogSchema",
             LogicalPlan::CreateCatalog(_create_catalog) => "CreateCatalog",
             LogicalPlan::CreateView(_create_view) => "CreateView",
-            LogicalPlan::SetVariable(_) => "SetVariable",
+            LogicalPlan::Statement(_) => "Statement",
+            // LogicalPlan::SetVariable(_) => "SetVariable",
```
[the remark this answers is not in the document]
Comment (jdye64) [reply dask_planner/src/sql/logical.rs:331]: `LogicalPlan::SetVariable` was removed in the latest version of DataFusion 22. I will just remove it instead of commenting it out.

File: dask_planner/src/sql/optimizer.rs (hunk 11 above)
Comment (jdye64) [reply dask_planner/src/sql/optimizer.rs:43]: TypeCorercion is no longer a OptimizerRule. Instead it was moved to be an Analyzer. Will just remove the commented out line entirely

File: dask_planner/.cargo/config.toml [hunk 13]
```diff
@@ -0,0 +1,11 @@
+[target.x86_64-apple-darwin]
```
[the remark this answers is not in the document]
Comment (jdye64) [reply dask_planner/.cargo/config.toml:1]: This is a fix for macOS builds. Specifically around substrait which ADP includes so therefore why it was added to this PR.

File: .github/workflows/rust.yml (hunk 4 above)
Comment (jdye64) [reply .github/workflows/rust.yml:37]: Actually this one needs to stay. Cargo Check runs a build command under the covers against ADP which in turn requires protoc to build out substrait proto files.

File: dask_sql/mappings.py (hunk 7 above)
Comment (jdye64) [reply dask_sql/mappings.py:185]: It is possible for `INTERVAL` or `INTERVAL_DAY` cases to be hit. They are still variants of DataFusion. I think we should leave those as are.

Agree on @sarahyurick logic though. I will make the change to make it assume 30 days for consistency sake.

File: dask_sql/physical/rex/core/literal.py (hunk 8 above)
Comment (jdye64) [reply dask_sql/physical/rex/core/literal.py:166]: It is possible so I think we leave it. Our tests just don't hit it.

File: tests/integration/test_rex.py [hunk 14]
```diff
@@ -926,8 +926,8 @@ def test_timestampdiff(c):
             "hours": [46, 6648, 591744],
             "days": [1, 277, 24656],
             "weeks": [0, 39, 3522],
-            "months": [0, 9, 810],
-            "quarters": [0, 3, 270],
+            "months": [0, 9, 821],
```
Comment (ayushdg) [inline tests/integration/test_rex.py:929]: Since this behavior is different from the results we get from mysql and a few implementations that support timestampdiff, I'm not sure whether we want to keep the existing logic or modify it. I didn't find an easy way to get a timestampdiff with datafusion/postgresql

**Review by ayushdg — APPROVED** [review]
Thanks a lot for these changes @jdye64!

File: pyproject.toml [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+[tool.isort]
+profile = "black"
+
+[tool.maturin]
+include = [
+    { path = "Cargo.lock", format = "sdist" }
+]
+exclude = [".github/**", "ci/**", ".asf.yaml"]
+# Require Cargo.lock is up to date
+locked = true
```
Comment (charlesbluca) [inline pyproject.toml:13]: Noting that I've been thinking about consolidating as much of our project metadata into pyproject.toml as possible (similar to other RAPIDS/Dask projects) which would make this file less stubby

**Review by charlesbluca — APPROVED** [review]

File: pyproject.toml (hunk 15 above)
Comment (ayushdg) [reply pyproject.toml:13]: +1 to this. I've also been thinking of simplifying the setup.py scripts and moving as much as possible into the pyproject.toml file
